### PR TITLE
strict name validation and --ignore-scripts flag

### PR DESCRIPTION
**Pull request by gronke**
Running the install command can have negative side-effects when installing a malicious package. This PR adds more explicit validation and allows to disable (un)install hook scripts.
### Fixes:
- #1886 
- #1887 
- #1888 
- #1889 
- #1890 
- #1891 
### Additional Features
- `--allow-symlinks` flag to explicitly allow `bower_components` to be a symlink
- `--disable-hooks` to disable all script hooks on systems that install assets from unknown sources

### CODE REVIEW

**Comment by gronke** [comment]
After finding a few behaviors that could infect deployment systems and local development setups as well, creating this patch was the first step before filing bugs related in the issue tracker.

It's not an easy to merge, because at some point new option flags had to be introduced and also checking package names too strict will break with existing packages in the wild. For example `d3` would become an invalid name, because it ends with a number. So before thinking of a merge, the consequences of such changes are TBD.

Thanks for your feedback!

**Comment by sheerun** [comment]
There are two blockers for this:

1. We can't introduce backward incompatibilities, so we need additional flag to enable strict validation of packages. We can introduce validation warnings, though.

2. It's probably better to put configuration into .bowerrc instead of command flags

**Comment by gronke** [comment]
1. I agree that breaking with existing repos is a bad idea (even with a major version bump this can confuse people). The least we can do is to provide flags that restrict Bower's behavior, so that nobody can execute code on your system. Let me quickly wrap up the issue sources and how I'd prefer to handle them

  1.a) illegal package names
  Package names should always be sanitized. Better tweak [the specification](https://github.com/bower/bower.json-spec) than allowing broken packages to exist. It can easily be determined which packages from the registry an bower update would break (just run the validator against all listed names). 
  One example package with illegal name is `d3` ("_Lowercase, a-z, can contain digits, 0-9, can contain dash or dot but not start/end with them._")

  1.b) symbolic links
  There's a good reason to make the target directory (`bower_components/`) a symlink, so that you can save disk space locally when working on a bunch of projects with overlapping dependencies. But on CI servers that can allow attacks using directory traversing. Disabling this **by default** is safe, since this is not documented as feature anywhere. Users who use the `--allow-symlinks` flag will accept the risk willingly.

  1.c) hook scripts
  In case you want to run `bower install` on a project you didn't review in detail (automated CI service scenario) it's safer to not allow hook script execution. That's what the `--disable-hooks` flag can be used. It's not enabled by default, but an additional feature that can be enabled when needed.

2. `.bowerrc` is not under your control under all circumstances. (e.g. you run a CI service that checks out arbitrary branches) We should not rely security features on files 3rd party.

--

The package names validation is the only thing that potentially can break existing packages and common install flow. What about adding deprecation warnings to all packages that don't follow the naming convention, preventing registration of new packages with invalid names at the Bower registry and use the time until the next major version release as grace period for malformed packages. The vulnerability could then be hot-patched with some blacklist-patterns for now.

**Comment by gronke** [comment]
Quite a few packages in the registry have an illegal name according to the specification. https://github.com/bower/json/pull/30 adds a unit test that runs `json.validate(package)` against all packages listed in the registry.

Test-Results: https://travis-ci.org/bower/json/jobs/82600244
```
  1) packages from bower registry should validate each listed package:
     Error: 2308/29181 package names do not validate
```

File: lib/commands/uninstall.js [hunk 1]
```diff
@@ -107,7 +110,9 @@ uninstall.readOptions = function (argv) {
 
     var options = cli.readOptions({
         'save': { type: Boolean, shorthand: 'S' },
-        'save-dev': { type: Boolean, shorthand: 'D' }
+        'save-dev': { type: Boolean, shorthand: 'D' },
+        'disable-hooks': { type: Boolean, shorthand: 'H' },
+        'allow-symlinks': { type: Boolean, shorthand: 'L' }
```
Comment (gronke) [inline lib/commands/uninstall.js]: Will remove this flag since #1891 was closed as `wontfix`

File: lib/commands/uninstall.js [hunk 2]
```diff
@@ -107,7 +110,9 @@ uninstall.readOptions = function (argv) {
 
     var options = cli.readOptions({
         'save': { type: Boolean, shorthand: 'S' },
-        'save-dev': { type: Boolean, shorthand: 'D' }
+        'save-dev': { type: Boolean, shorthand: 'D' },
+        'disable-hooks': { type: Boolean, shorthand: 'H' },
```
Comment (sheerun) [inline lib/commands/uninstall.js]: Can you also change this options to --ignore-scripts, just like NPM?

**Comment by gronke** [comment]
@sheerun changed the flag to `--ignore-scripts` (like npm also does). Did you see https://github.com/bower/json/pulls/gronke? One of the PR's fails the test, but prints a list of illegal package names that are already in the registry. I suggest to tweak the specification, so that their names become valid subsequently.

**Comment by faceleg** [comment]
@gronke please can you update your branch to make it mergeable?

@sheerun how does this look to you now gronke has changed the flag?

@bower/bower-volunteers additional reviewers needed please

**Comment by gronke** [comment]
Will rebase shortly. Thanks @faceleg for looking into this!

**Comment by sheerun** [comment]
@faceleg Can you install packages with invalid name with this PR?

**Comment by gronke** [comment]
@sheerun https://github.com/bower/json/pull/29/files#diff-3f9d46363ce3a43200a9dfa6e3c9cf28R178 does a minimum check of the package name for consisting of valid characters only. That does not enforce package name patterns from the bowerJson specification, but prevents directory traversals.

**Comment by sheerun** [comment]
Thank you for effort but bower has been depreacted and we point to Yarn instead. Also this honestly this seems like minor security concern as pre- and postinstall scripts can be defined only on app level so app developers would need to sabotage themselves to perform this kind of attack, and there's easier way to perform this attack: just put some script in application code itself.

**Pull request closed without merging**